cmd/kube-controller-manager/app/options/options_test.go

```diff
@@ -657,6 +657,104 @@ func TestApplyTo(t *testing.T) {
 	}
 }
 
+// TODO: this test having some conflict on feature gate:
+// panic: cannot add a feature gate after adding it to the flag set [recovered]
+func TestEmulatedVersion(t *testing.T) {
+	var cleanupAndSetupFunc = func() featuregate.FeatureGate {
+		componentGlobalsRegistry := utilversion.DefaultComponentGlobalsRegistry
+		componentGlobalsRegistry.Reset() // make sure this test have a clean state
+		t.Cleanup(func() {
+			componentGlobalsRegistry.Reset() // make sure this test doesn't leak a dirty state
+		})
+
+		verKube := utilversion.NewEffectiveVersion("1.32")
+		fg := utilfeature.DefaultFeatureGate.DeepCopy()
+		utilruntime.Must(fg.AddVersioned(map[featuregate.Feature]featuregate.VersionedSpecs{
+			"kubeA": {
+				{Version: version.MustParse("1.32"), Default: true, LockToDefault: true, PreRelease: featuregate.GA},
+				{Version: version.MustParse("1.30"), Default: false, PreRelease: featuregate.Beta},
+			},
+			"kubeB": {
+				{Version: version.MustParse("1.31"), Default: false, PreRelease: featuregate.Alpha},
+			},
+		}))
+		utilruntime.Must(componentGlobalsRegistry.Register(utilversion.DefaultKubeComponent, verKube, fg))
+		return fg
+	}
+
+	testcases := []struct {
+		name              string
+		flags             []string
+		wantErr           bool
+		errorSubString    string
+		wantFeaturesGates map[string]bool
+	}{
+		{
+			name:              "default feature gates at binary version",
+			flags:             []string{},
+			wantErr:           false,
+			wantFeaturesGates: map[string]bool{"kubeA": true, "kubeB": false},
+		},
+		{
+			name: "emulating version out of range",
+			flags: []string{
+				"--emulated-version=1.28",
+			},
+			wantErr:           false, // TODO, expecting error.
```

Comment:
@siyuanfoundation  also need your help here as well, some how it didn't trigger the validation. from what I debug, it seems the env is already shifted to 1.32, it dodged the ValidateKubeEffectiveVersion with condition != 1.31.

Reply:
I did merge master, maybe that's the issue? I don't see 1.31 branch, is there some branch I should rebase to?